### PR TITLE
mkmf.rb: sort lists of header files

### PR DESCRIPTION
**Pull request by zimbatm**
This is a continuation of 15dba481a28164aea17dc59f4026e0c1bb733772 /
GH-1367.

Without sorting the list of headers explicitly, its order is
indeterministic, because readdir() is also not deterministic. When the
list of object files varies between builds, it generates a different
Makefile, which results in an unreproducible build.

Diffoscope example: https://r13y.com/diff/8350ec80865096f84705f3e000d97f151760e768bf626e52c84570413c5e1093-f9dea788fe62d683e90e12e166f29b778c39c699a5be39f9262e5cffb77d3e45.html

### CODE REVIEW

**Comment by zimbatm** [comment]
I also pushed another patch to make the sitearchdir and sitelibdir in the generated Makefile reproducible.

At this point, the patchset needs a bit more testing. I would be grateful for high-level feedback.

**Comment by zimbatm** [comment]
I ended-up removing the Makefile from the generated output so those patches are not strictly necessary for me anymore: https://github.com/NixOS/nixpkgs/pull/101655

**Comment by nobu** [comment]
`Dir.[]` sorts the result in the master.

File: lib/rubygems/ext/ext_conf_builder.rb [hunk 1]
```diff
@@ -12,7 +12,13 @@ def self.build(extension, dest_path, results, args=[], lib_dir=nil)
     require 'fileutils'
     require 'tempfile'
 
-    tmp_dest = Dir.mktmpdir(".gem.", ".")
+    # SOURCE_EPOCH_DATE is used for reproducible builds. Assume that we are
+    # in a build sandbox when it's set.
+    if ENV['SOURCE_EPOCH_DATE']
+      tmp_dest = Time.at(ENV['SOURCE_EPOCH_DATE'].to_i).strftime(".gem.%Y%m%d")
```
Comment (nobu) [inline lib/rubygems/ext/ext_conf_builder.rb:18]: No `mkdir`?

Comment (zimbatm) [reply lib/rubygems/ext/ext_conf_builder.rb:18]: good catch, fixed

**Comment by zimbatm** [comment]
I don't know about `Dir.[]`. Wouldn't it make it harder to back-port the patch? The file is already using `Dir[].sort` in one other place.

**Pull request closed without merging**